--- ./include/net/sock.h	2002/02/06 15:25:10	1.1
+++ ./include/net/sock.h	2002/05/22 12:14:56
@@ -488,7 +488,13 @@
 	} bictcp;
 };

-
+#if 1
+#define UDP_OPT_IN_SOCK 1
+struct udp_opt {
+	__u32 esp_in_udp;
+};
+#endif
+
 /*
  * This structure really needs to be cleaned up.
  * Most of it is for TCP, and not used by any of
@@ -655,6 +661,9 @@
 #if defined(CONFIG_SPX) || defined (CONFIG_SPX_MODULE)
 		struct spx_opt		af_spx;
 #endif /* CONFIG_SPX */
+#if 1
+		struct udp_opt          af_udp;
+#endif

 	} tp_pinfo;

